### PR TITLE
Optimize DefaultRequestDispatcher and improve ssh-agent warning message

### PR DESCRIPTION
**Pull request by davidxia**
and only get them if an endpoint is HTTPS.
We were asking the ssh-agent for the list of identities
for every new connection to an endpoint.

### CODE REVIEW

**Comment by mattnworb** [comment]
Is the `Identity` instance usable after the `AgentProxy` has been closed?

**Comment by mattnworb** [comment]
Looks like it is usable, so nevermind.

**Comment by mattnworb** [comment]
:+1: to the change itself.

I think this class would be a lot more testable though (it has no test today) if instead you passed either the `List<Identity>` or `AgentProxy` instance in the constructor, so that the class that is making http requests didn't also have to be concerned with how to get the ssh-agent identities or how to get an AgentProxy etc.

**Comment by davidxia** [comment]
Yea, I haven't had the chance to add a test yet. Good idea about passing in
the list of identities. But it seems like if you don't create a new
instance of an AgentProxy on every connection, the TLS handshake is messed
up somehow. I wasn't able to drill down into why though...

On Fri, Nov 20, 2015 at 1:09 PM, Matt Brown <notifications@github.com>
wrote:

> [image: :+1:] to the change itself.
>
> I think this class would be a lot more testable though (it has no test
> today) if instead you passed either the List<Identity> or AgentProxy
> instance in the constructor, so that the class that is making http requests
> didn't also have to be concerned with how to get the ssh-agent identities
> or how to get an AgentProxy etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/helios/pull/757#issuecomment-158478315>.
>



-- 
David Xia
Software Engineer
dxia@spotify.com

File: helios-client/src/main/java/com/spotify/helios/client/DefaultRequestDispatcher.java [hunk 1]
```diff
@@ -83,6 +85,22 @@
     endpointIterator = EndpointIterator.of(endpoints);
     this.executorService = executorService;
     this.user = user;
+
+    if (endpointIterator.hasHttps()) {
+      try {
+        identities = getSshIdentities();
+      } catch (Exception e) {
+        log.debug("Couldn't get identities from ssh-agent", e);
+        log.warn(
```
Comment (davidxia) [inline helios-client/src/main/java/com/spotify/helios/client/DefaultRequestDispatcher.java]: @gimaker @mattnworb This should now happen only once. Open to suggestions on how to improve this message.

**Comment by davidxia** [comment]
I'm having trouble thinking of a sane way to make `DefaultRequestDispatcher` more testable. How could I mock out the URL connection part?...

**Pull request closed without merging**